File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        minBuy = [prices[0]] * len(prices)
        for i in range(1, len(prices)):
            minBuy[i] = min(minBuy[i - 1], prices[i - 1])
        
        for i in range(1, len(prices)):
            res = max(res, prices[i] - minBuy[i])
        return res

File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-1.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        return Rec(prices, 0, false);
    }

    private int Rec(int[] prices, int i, bool bought) {
        if (i == prices.Length) {
            return 0;
        }

        int res = Rec(prices, i + 1, bought);

        if (bought) {
            res = Math.Max(res, prices[i] + Rec(prices, i + 1, false));
        } else {
            res = Math.Max(res, -prices[i] + Rec(prices, i + 1, true));
        }

        return res;
    }
}